# 作業 : (Kaggle)鐵達尼生存預測
https://www.kaggle.com/c/titanic

# [作業目標]
- 試著調整特徵篩選的門檻值, 觀察會有什麼影響效果

# [作業重點]
- 調整相關係數過濾法的篩選門檻, 看看篩選結果的影響 (In[5]~In[8], Out[5]~Out[8])
- 調整L1 嵌入法篩選門檻, 看看篩選結果的影響 (In[9]~In[11], Out[9]~Out[11])

In [1]:
# 做完特徵工程前的所有準備 (與前範例相同)
import pandas as pd
import numpy as np
import copy
from sklearn.preprocessing import LabelEncoder, MinMaxScaler
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression

data_path = './D22_DATA/'
df = pd.read_csv(data_path + 'titanic_train.csv')

train_Y = df['Survived']
df = df.drop(['PassengerId'] , axis=1)
df.head()

Survived  Pclass                                               Name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S  
1  female  38.0      1      0          PC 17599  71.2833   C85        C  
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3  female  35.0      1      0            113803  53.1000  C123        S  
4    male  35.0      0      0            373450   8.0500   NaN        S

In [2]:
# 計算df整體相關係數, 並繪製成熱圖
import seaborn as sns
import matplotlib.pyplot as plt
corr = df.corr()
sns.heatmap(corr)
plt.show()

<Figure size 640x480 with 2 Axes>

In [3]:
corr

Survived    Pclass       Age     SibSp     Parch      Fare
Survived  1.000000 -0.338481 -0.077221 -0.035322  0.081629  0.257307
Pclass   -0.338481  1.000000 -0.369226  0.083081  0.018443 -0.549500
Age      -0.077221 -0.369226  1.000000 -0.308247 -0.189119  0.096067
SibSp    -0.035322  0.083081 -0.308247  1.000000  0.414838  0.159651
Parch     0.081629  0.018443 -0.189119  0.414838  1.000000  0.216225
Fare      0.257307 -0.549500  0.096067  0.159651  0.216225  1.000000

In [4]:
# 記得刪除 Survived
df = df.drop(['Survived'] , axis=1)

#只取 int64, float64 兩種數值型欄位, 存於 num_features 中
num_features = []
for dtype, feature in zip(df.dtypes, df.columns):
    if dtype == 'float64' or dtype == 'int64':
        num_features.append(feature)
print(f'{len(num_features)} Numeric Features : {num_features}\n')

# 削減文字型欄位, 只剩數值型欄位
df = df[num_features]
df = df.fillna(-1)
MMEncoder = MinMaxScaler()
df.head()

5 Numeric Features : ['Pclass', 'Age', 'SibSp', 'Parch', 'Fare']



Pclass   Age  SibSp  Parch     Fare
0       3  22.0      1      0   7.2500
1       1  38.0      1      0  71.2833
2       3  26.0      0      0   7.9250
3       1  35.0      1      0  53.1000
4       3  35.0      0      0   8.0500

# 作業1
* 鐵達尼生存率預測中，試著變更兩種以上的相關係數門檻值，觀察預測能力是否提升?

In [5]:
# 原始特徵 + 邏輯斯迴歸
train_X = MMEncoder.fit_transform(df)
estimator = LogisticRegression(solver='liblinear')
cross_val_score(estimator, train_X, train_Y, cv=5).mean()

0.7038635542329971

In [6]:
# 篩選相關係數1
high_list = list(corr.loc[(corr['Survived']>0.1) | (corr['Survived']<-0.1), 'Survived'].index)
high_list.pop(0)
print(high_list)

['Pclass', 'Fare']


In [7]:
# 特徵1 + 邏輯斯迴歸
train_X = MMEncoder.fit_transform(df[high_list])
cross_val_score(estimator, train_X, train_Y, cv=5).mean()

0.6791567235397566

In [8]:
# 篩選相關係數2
"""
Your Code Here
"""

high_list = list(corr.loc[(corr['Survived']>0.05) | (corr['Survived']<-0.05), 'Survived'].index)
high_list.pop(0)
print(high_list)

['Pclass', 'Age', 'Parch', 'Fare']


In [9]:
# 特徵2 + 邏輯斯迴歸
train_X = MMEncoder.fit_transform(df[high_list])
cross_val_score(estimator, train_X, train_Y, cv=5).mean()

0.7049872206659582

### [作業1] 鐵達尼生存率預測中，試著變更兩種以上的相關係數門檻值，觀察預測能力是否提升?
***
[Ans] 使用不同相關係數門檻，確實可以會達到不同的預測能力。
      - 相關係數設定為: >0.1 、 <-0.1 , 準確率只達0.679
      - 相關係數設定為: >0.05、 <-0.05, 準確率可提升至0.705
      
[心得] 
    - 使用較高的門檻值可能會過濾太多特徵，而導致預測上的準確率不理想。
    - 當可以參考的特徵本來就不多時，又過濾了較低相關性的特徵，這麼一來就會影響準確率。
    - 因此，門檻值的大小將會影響特徵留下的多寡，也同時影響預測的準確率。


# 作業2
* 續上題，使用 L1 Embedding 做特徵選擇(自訂門檻)，觀察預測能力是否提升?

In [10]:
df

Pclass   Age  SibSp  Parch     Fare
0         3  22.0      1      0   7.2500
1         1  38.0      1      0  71.2833
2         3  26.0      0      0   7.9250
3         1  35.0      1      0  53.1000
4         3  35.0      0      0   8.0500
..      ...   ...    ...    ...      ...
886       2  27.0      0      0  13.0000
887       1  19.0      0      0  30.0000
888       3  -1.0      1      2  23.4500
889       1  26.0      0      0  30.0000
890       3  32.0      0      0   7.7500

[891 rows x 5 columns]

In [11]:
from sklearn.linear_model import Lasso
"""
Your Code Here, select parameter alpha 
"""
from sklearn.linear_model import Lasso

L1_Reg = Lasso(alpha=0.005)
train_X = MMEncoder.fit_transform(df)
L1_Reg.fit(train_X, train_Y)
L1_Reg.coef_

array([-0.39126312, -0.1355752 , -0.        ,  0.03068895,  0.        ])

In [12]:
from itertools import compress
L1_mask = list((L1_Reg.coef_>0) | (L1_Reg.coef_<0))
L1_list = list(compress(list(df), list(L1_mask)))
L1_list

['Pclass', 'Age', 'Parch']

In [13]:
# L1_Embedding 特徵 + 線性迴歸
train_X = MMEncoder.fit_transform(df[L1_list])
cross_val_score(estimator, train_X, train_Y, cv=5).mean()

0.6993628951406613

### [作業2] 續上題，使用 L1 Embedding 做特徵選擇(自訂門檻)，觀察預測能力是否提升?
***
[Ans] 使用L1 Embedding預測能力的狀況較不穩定。

      1. 原特徵 + LogisticRegression : 0.704
      2. 篩選後特徵 + LogisticRegression : 0.705
      3. L1 Embedding + LogisticRegression : 0.704(alpha=0.001), 0.699(alpha=0.005)
      
[心得]      
      - 當我們設定Lasso()中的alpha值時，它會影響正規化程度，呈現的結果就是這個特徵的重要性。
      - 另外還需在設立門檻值來篩選掉正規化後數值較小的特徵。
      - 而這邊設立門檻值又回到了剛剛第2點，門檻值決定保留哪些特徵。